### DISCUSSION TITLE
Parameter writings / EEPROM access / MaxChargeCurrent

### DISCUSSION THREAD

**Discussion by beaune33** [PatrikTrestik/homeassistant-solax-http#21] [Ideas]
I am wondering if we have any information, which parameters are stored in an EEPROM and how often they may be written. For the inverter we have the well-known document, where parameters are marked with a star, if they are stored in the EEPROM, and therefore should not be written to often. Are there any comparable information for the EVC?

In my concrete case I am thinking about changing MaxChargeCurrent (Modbus 0x628) in Fast Charging mode dynamically, depending on the energy available. Unfortunately my car starts charging not beyond 8A, so I cannot use green mode and have to program something similar on my own. So, I could start with 8A in the morning when sun is low, if sun increases and more power is available, I could dynamically switch to 9/10A etc, and if sun is going down in the same way. The only thing is, I am afraid that this value could be written too often and therefore the life time of the EEPROM could decrease. But the question is, if MaxChargeCurrent  is stored in EPROM at all, or if this is a volatile setting. Do you have any information about this?

**Comment by Philipp-X1** [comment]
May I ask how you are going to realise the dynamic charging? All programming by hand?
Do you know about EVCC? It could do the job perfectly and is already here. So we would not need to reinvent the wheel. 

At the moment I am trying to set up a custom charger at EVCC but I am struggling with the Modbus settings… Is there anybody here interested in Setting up EVCC and would want to help?
(Of course EVCC could be Setup with HTTP as well)

**Reply by lieglw** [reply]
Hi @Philipp-X1  I also tried to run EVCC, but the EV Charger I never got in. Had the inverter running with the WiFi dongle and for HA it is stable.
Did a Modbus Proxy for the EVCC for the inverter.
Can you help me with the Solax Charger?

**Reply by Philipp-X1** [reply]
Yes, but this is the wrong place here.  Probably best to do it over slack.

**Reply by lieglw** [reply]
Hi, yes please, but do not know how.

**Comment by beaune33** [comment]
Everything by hand, yes. I have all necessary information in my local network accessible by home automation, inverter, evc,  access to car via OBD interface,… so it is easy to program the rest in fhem without other external  software. I am not sure how evcc could help or make things more easy. At the end the question is the same: the charging energy needs to be adjusted, so I need to communicate to the evc. The way to do seems to be the parameter mentioned above. I know it works, the only thing I am asking myself is if I have to be careful because of eeprom writing cycles. Maybe not, but I am not sure.

**Comment by Philipp-X1** [comment]
Well at least I was reading somewhere that the Inverter ( Solax X3 Hybrid ) has 1.000.000 eeprom writing cycles. I don’t know how it is realized with the EV Charger. 

EVCC has a good UI and can do quite more than the integrated smart function from Solax. 

I got all the information in my Network aswell and I also changed the charge rate already. But I can’t write the software :)

**Comment by PatrikTrestik** [comment]
EVC Modbus document I have does not show any information about EPROM written registers. But there is same disclaimer as in inverter document so I expect that some exist.
Better to ask Solax support. 
1M writes is standard EEPROM cycle limit.

**Comment by PatrikTrestik** [comment]
> I am wondering if we have any information, which parameters are stored in an EEPROM and how often they may be written. For the inverter we have the well-known document, where parameters are marked with a star, if they are stored in the EEPROM, and therefore should not be written to often. Are there any comparable information for the EVC?
> 
> In my concrete case I am thinking about changing MaxChargeCurrent (Modbus 0x628) in Fast Charging mode dynamically, depending on the energy available. Unfortunately my car starts charging not beyond 8A, so I cannot use green mode and have to program something similar on my own. So, I could start with 8A in the morning when sun is low, if sun increases and more power is available, I could dynamically switch to 9/10A etc, and if sun is going down in the same way. The only thing is, I am afraid that this value could be written too often and therefore the life time of the EEPROM could decrease. But the question is, if MaxChargeCurrent is stored in EPROM at all, or if this is a volatile setting. Do you have any information about this?

Your idea is probably better solvable with simulating Green function of SolaX inverter. It periodically (3s) writes "Available PV power" to EVC if connected.
See here: https://github.com/wills106/homeassistant-solax-modbus/discussions/666

**Comment by beaune33** [comment]
Interesting idea... but I am not completely sure how it could work. My problem is: my car needs at least 8A for charging. For green mode I can only configure 3 or 6A, so this does not work at all. For ECO mode I can configure 6 or 10A. 6A would also do nothing, so I only have 10A. But for normal sunny days this is too much. I would need a green or ECO mode for 8A, that would be ideal.

Because both modes do not offer 8A, I have to go to Fast mode. There I can configure the max. current to 8A.

What could be achieved by writing "Available PV power" to EVC? Of course, I could cut the direct modbus connection and send the value by my home automation. Or just cyclically overwrite them. But would this really help? I would still need fast mode I guess. Maybe I did not get your idea correctly.

Regarding the EVC Modbus document you mentioned above: Is this available somewhere? Could you share it? Till now I have all information from your source code only.

**Reply by romankysely** [reply]
Modbus document: https://gbc-solino.cz/wp-content/uploads/2022/07/Hybrid-X1X3-G4-ModbusTCPRTU-V3.21-English_0622-public-version.pdf

**Comment by PatrikTrestik** [comment]
In EVC Modbus document there is nothing marked as EEPROM (*), but it is rather mistake than reality.
Most of "setup" registers are stored in EEPROM as we know from Inverter documentation. My guess is that MaxChargePower will be EEPROM too.

Better solution is to write to reg 0x701 and following. Its meaning is "Available Green power" but you can write what you want. 
If you need charge curret 8A: 
1) write 1800W and EVC will start charging in Green
2) Read actual EVC charging current C. X=8-C.
3) Write X*230 to  0x701. EVC will stay on 8A current.

Registers 0x700+ are definitely safe, because Inverter writes to it 1 per second.
Only downside is that no documentation exists and you have to believe my investigation or do your own tests.

BTW:
If you are going to program this in high level (say computer with HA) you are violating first requirement of good regulation loop. Inner/critical loop should be as simple as possible and using dedicated hardware. ESP32 component will be better.
More complicated software or hardware means more possibility for crash :)

**Comment by Philipp-X1** [comment]
I was chatting with Solax support yesterday. 
Asking them: _Is this 0x0628 register written in EEPROM?_

![IMG_2976](https://github.com/PatrikTrestik/homeassistant-solax-http/assets/146736973/80e0d76e-2091-4230-a7c3-4c743d611a64)

The answer I got:

_Thank you for your patience. 0x0628 doesn't write to EEPROM. It's for parallel._

I don’t know what he means with parallel, but it seems like changing 0x0628 seems to be ok.

If we need more info about the registers and eeprom we could try to contact Solax, seems like they are willing to help.

**Reply by beaune33** [reply]
Perfect, thanks for this information. I like the idea of changing settings like this on special events more than influencing process values, especially because of the regulation loop topic.

**Reply by lieglw** [reply]
Hi, could that be for the parallel mode when you have 2 EV Chargers?

**Comment by empx92** [comment]
@Philipp-X1 did you already get it to work in EVCC over http? Patriks http integration in HA works just fine but the modbus in EVCC not...

**Comment by Philipp-X1** [comment]
Nope, no need for me. I am using the modbus connection.
